File: app/src/main/java/com/past/music/adapter/MyContentAdapter.java
package com.past.music.adapter;/**
 * Created by gaojin on 2017/1/26.
 */

import android.content.Context;
import android.content.Intent;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.facebook.drawee.view.SimpleDraweeView;
import com.neu.gaojin.MyOkHttpClient;
import com.neu.gaojin.response.BaseCallback;
import com.past.music.MyApplication;
import com.past.music.activity.CollectedActivity;
import com.past.music.activity.KtBaseActivity;
import com.past.music.activity.RecentMusicActivity;
import com.past.music.activity.SongListActivity;
import com.past.music.activity.SongListInfoActivity;
import com.past.music.api.AvatarRequest;
import com.past.music.api.AvatarResponse;
import com.past.music.entity.MusicEntity;
import com.past.music.entity.SongListEntity;
import com.past.music.fragment.KtDownLoadFragment;
import com.past.music.fragment.KtLocalMusicFragment;
import com.past.music.pastmusic.R;
import com.past.music.utils.MConstants;
import com.past.music.utils.MusicUtils;
import com.past.music.widget.CircleImageView;
import com.past.music.widget.MineItemView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * =======================================================
 * 作者：gaojin
 * 日期：2017/3/4 下午3:17
 * 描述：我的Fragment界面的adapter
 * 备注：Copyright © 2010-2017. gaojin All rights reserved.
 * =======================================================
 */
public class MyContentAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int HEAD_LAYOUT = 0X01;
    private static final int FUNC_LAYOUT = 0X02;
    private static final int FAVOR_TITLE_LAYOUT = 0X03;
    private static final int FAVOR_ITEM_LAYOUT = 0X04;
    private static final int MANAGE_SONG_LIST_LAYOUT = 0X05;

    private List<SongListEntity> mList;
    private Context mContext = null;

    public MyContentAdapter(Context context) {
        this.mContext = context;
        mList = MyApplication.songListDBService.query();
    }

    public void setSongList(List<SongListEntity> listItem) {
        this.mList = listItem;
        notifyItemRangeChanged(3, listItem.size() + 1);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == HEAD_LAYOUT) {
            return new HeadLayoutHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.mine_head_layout, parent, false));
        } else if (viewType == FUNC_LAYOUT) {
            return new FuncLayoutHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.mine_func_layout, parent, false));
        } else if (viewType == FAVOR_TITLE_LAYOUT) {
            return new FavorTitleLayout(LayoutInflater.from(parent.getContext()).inflate(R.layout.mine_favor_title_layout, parent, false));
        } else if (viewType == MANAGE_SONG_LIST_LAYOUT) {
            return new ManageSongList(LayoutInflater.from(parent.getContext()).inflate(R.layout.manage_songlist_layout, parent, false));
        } else {
            return new FavorItemLayout(LayoutInflater.from(parent.getContext()).inflate(R.layout.mine_favor_item_layout, parent, false));
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof HeadLayoutHolder) {

        } else if (holder instanceof FuncLayoutHolder) {
            ((FuncLayoutHolder) holder).onBind();
        } else if (holder instanceof FavorTitleLayout) {

        } else if (holder instanceof FavorItemLayout) {
            ((FavorItemLayout) holder).onBind(mList.get(position - 3));
        }
    }

    @Override
    public int getItemCount() {
        if (mList == null) {
            return 4;
        } else {
            return mList.size() + 4;
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return HEAD_LAYOUT;
        } else if (position == 1) {
            return FUNC_LAYOUT;
        } else if (position == 2) {
            return FAVOR_TITLE_LAYOUT;
        } else {
            if (position == mList.size() + 3) {
                return MANAGE_SONG_LIST_LAYOUT;
            } else {
                return FAVOR_ITEM_LAYOUT;
            }
        }
    }

    class HeadLayoutHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.mine_head_avatar)
        CircleImageView mHeadAvatar;

        @BindView(R.id.mine_head_listen_time)
        TextView mListenTime;

        @BindView(R.id.mine_head_listen_vip)
        TextView mListenVip;

        @BindView(R.id.mine_head_user_name)
        TextView mUserName;

        public HeadLayoutHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    class FuncLayoutHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.local_music)
        MineItemView mLocalMusic;

        @BindView(R.id.download_music)
        MineItemView mDownMusic;

        @BindView(R.id.recent_music)
        MineItemView mRecentMusic;

        @BindView(R.id.love_music)
        MineItemView mLoveMusic;

        @BindView(R.id.love_singer)
        MineItemView mLoveSinger;

        @BindView(R.id.buy_music)
        MineItemView mBuyMusic;

        @OnClick(R.id.local_music)
        void localMusic() {
            FragmentTransaction ft = ((KtBaseActivity) mContext).getSupportFragmentManager().beginTransaction();
            ft.addToBackStack(null);
            ft.setCustomAnimations(R.anim.slide_right_in, R.anim.slide_left_out, R.anim.slide_left_in, R.anim.slide_right_out);
            ft.add(R.id.content, KtLocalMusicFragment.Companion.newInstance(0)).commitAllowingStateLoss();

        }

        @OnClick(R.id.download_music)
        void downloadMusic() {
            FragmentTransaction ft = ((KtBaseActivity) mContext).getSupportFragmentManager().beginTransaction();
            ft.addToBackStack(null);
            ft.setCustomAnimations(R.anim.slide_right_in, R.anim.slide_left_out, R.anim.slide_left_in, R.anim.slide_right_out);
            ft.add(R.id.content, KtDownLoadFragment.Companion.newInstance()).commitAllowingStateLoss();
        }

        @OnClick(R.id.recent_music)
        void recentMusic() {
            Intent intent = new Intent(mContext, RecentMusicActivity.class);
            ((KtBaseActivity) mContext).startActivityByX(intent, true);
        }

        @OnClick(R.id.love_music)
        void loveMusic() {
            Intent intent = new Intent(mContext, CollectedActivity.class);
            ((KtBaseActivity) mContext).startActivityByX(intent, true);
        }

        @OnClick(R.id.love_singer)
        void loveSinger() {
            FragmentTransaction ft = ((KtBaseActivity) mContext).getSupportFragmentManager().beginTransaction();
            ft.addToBackStack("KtLocalMusicFragment");
            ft.setCustomAnimations(R.anim.slide_right_in, R.anim.slide_left_out, R.anim.slide_left_in, R.anim.slide_right_out);
            ft.add(R.id.content, KtLocalMusicFragment.Companion.newInstance(0)).commitAllowingStateLoss();
        }

        @OnClick(R.id.buy_music)
        void buyMusic() {
            Toast.makeText(mContext, "测试", Toast.LENGTH_SHORT).show();
        }

        public FuncLayoutHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        public void onBind() {
            int localCount = MusicUtils.queryMusic(mContext, MConstants.START_FROM_LOCAL).size();
//            int singerCount = MusicUtils.queryArtist(mContext).size();
//            int albumCount = MusicUtils.queryAlbums(mContext).size();
//            int folderCount = MusicUtils.queryFolder(mContext).size();

            mLocalMusic.setmItemCount(String.valueOf(localCount));

        }
    }

    class FavorTitleLayout extends RecyclerView.ViewHolder {

        @BindView(R.id.rl_favor_title)
        RelativeLayout mTitleLayout;

        @BindView(R.id.playlist_count)
        TextView mPlayListCount;

        @OnClick(R.id.rl_favor_title)
        void favorTltle() {
            Intent intent = new Intent(mContext, SongListActivity.class);
            ((KtBaseActivity) mContext).startActivityByX(intent, false);
        }

        public FavorTitleLayout(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    class FavorItemLayout extends RecyclerView.ViewHolder {

        @BindView(R.id.img_album)
        SimpleDraweeView songList_pic;

        @BindView(R.id.play_list_title)
        TextView mPlayListTitle;

        @BindView(R.id.play_list_info)
        TextView mPlayListInfo;

        @BindView(R.id.img_down)
        ImageView mImgDown;

        @BindView(R.id.rl_favor_item)
        RelativeLayout mItemLayout;

        @OnClick(R.id.rl_favor_item)
        void favorItem() {
            SongListInfoActivity.startActivity(mContext, mList.get(getAdapterPosition() - 3).getId(), mList.get(getAdapterPosition() - 3).getName());
        }

        public FavorItemLayout(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        public void onBind(final SongListEntity songListEntity) {
            final MusicEntity musicEntity = MyApplication.musicInfoDBService.firstEntity(songListEntity.getId());
            if (musicEntity != null && musicEntity.getAlbumPic() != null) {
                MyApplication.songListDBService.updatePic(songListEntity.getId(), musicEntity.getAlbumPic());
                songList_pic.setImageURI(musicEntity.getAlbumPic());
            } else if (musicEntity != null) {
                if (MyApplication.imageDBService.query(musicEntity.getArtist().replace(";", "")) == null) {
                    AvatarRequest avatarRequest = new AvatarRequest();
                    avatarRequest.setArtist(musicEntity.getArtist().replace(";", ""));
                    MyOkHttpClient.getInstance(mContext).sendNet(avatarRequest, new BaseCallback<AvatarResponse>() {
                        @Override
                        public void onFailure(int code, String error_msg) {

                        }

                        @Override
                        public void onSuccess(int statusCode, final AvatarResponse response) {
                            MyApplication.imageDBService.insert(musicEntity.getArtist().replace(";", ""), response.getArtist().getImage().get(2).get_$Text112());
                            MyApplication.songListDBService.updatePic(songListEntity.getId(), response.getArtist().getImage().get(3).get_$Text112());
                            songList_pic.setImageURI(response.getArtist().getImage().get(2).get_$Text112());
                        }
                    });
                } else {
                    MyApplication.songListDBService.updatePic(songListEntity.getId(), MyApplication.imageDBService.query(musicEntity.getArtist().replace(";", "")));
                    songList_pic.setImageURI(MyApplication.imageDBService.query(musicEntity.getArtist().replace(";", "")));
                }
            }
            mPlayListTitle.setText(songListEntity.getName());
            mPlayListInfo.setText(MyApplication.musicInfoDBService.getLocalCount(songListEntity.getId()));
        }
    }

    class ManageSongList extends RecyclerView.ViewHolder implements View.OnClickListener {
        public ManageSongList(View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            Intent intent = new Intent(mContext, SongListActivity.class);
            ((KtBaseActivity) mContext).startActivityByX(intent, false);
        }
    }
}

File: app/src/main/java/com/past/music/adapter/MyContentAdapter.kt
package com.past.music.adapter

import android.content.Context
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.Toast
import com.facebook.drawee.view.SimpleDraweeView
import com.neu.gaojin.MyOkHttpClient
import com.neu.gaojin.response.BaseCallback
import com.past.music.MyApplication
import com.past.music.activity.BaseActivity
import com.past.music.activity.CollectedActivity
import com.past.music.activity.SongListActivity
import com.past.music.activity.SongListInfoActivity
import com.past.music.api.AvatarRequest
import com.past.music.api.AvatarResponse
import com.past.music.entity.SongListEntity
import com.past.music.fragment.DownLoadFragment
import com.past.music.fragment.LocalMusicFragment
import com.past.music.fragment.RecentMusicFragment
import com.past.music.pastmusic.R
import com.past.music.utils.MConstants
import com.past.music.utils.MusicUtils
import com.past.music.widget.CircleImageView
import com.past.music.widget.MineItemView

/**
 * Created by gaojin on 2017/12/7.
 */
class MyContentAdapter constructor(context: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val HEAD_LAYOUT = 0X01
    private val FUNC_LAYOUT = 0X02
    private val FAVOR_TITLE_LAYOUT = 0X03
    private val FAVOR_ITEM_LAYOUT = 0X04
    private val MANAGE_SONG_LIST_LAYOUT = 0X05

    private val START = 3
    private var mList: List<SongListEntity>? = null
    private var mContext: Context = context

    init {
        mList = MyApplication.songListDBService.query()
    }

    fun setSongList(listItem: List<SongListEntity>) {
        this.mList = listItem
        notifyItemRangeChanged(START, listItem.size + 1)
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            HEAD_LAYOUT -> HeadLayoutHolder(LayoutInflater.from(parent!!.context).inflate(R.layout.mine_head_layout, parent, false))
            FUNC_LAYOUT -> FuncLayoutHolder(LayoutInflater.from(parent!!.context).inflate(R.layout.mine_func_layout, parent, false))
            FAVOR_TITLE_LAYOUT -> FavorTitleLayout(LayoutInflater.from(parent!!.context).inflate(R.layout.mine_favor_title_layout, parent, false))
            MANAGE_SONG_LIST_LAYOUT -> ManageSongList(LayoutInflater.from(parent!!.context).inflate(R.layout.manage_songlist_layout, parent, false))
            else -> FavorItemLayout(LayoutInflater.from(parent!!.context).inflate(R.layout.mine_favor_item_layout, parent, false))
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder?, position: Int) {
        if (holder is HeadLayoutHolder) {

        } else (holder as? FuncLayoutHolder)?.onBind() ?: if (holder is FavorTitleLayout) {

        } else (holder as? FavorItemLayout)?.onBind(mList!![position - 3])
    }

    override fun getItemCount(): Int {
        return if (mList == null) {
            4
        } else {
            mList!!.size + 4
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) {
            HEAD_LAYOUT
        } else if (position == 1) {
            FUNC_LAYOUT
        } else if (position == 2) {
            FAVOR_TITLE_LAYOUT
        } else {
            if (position == mList!!.size + 3) {
                MANAGE_SONG_LIST_LAYOUT
            } else {
                FAVOR_ITEM_LAYOUT
            }
        }
    }

    internal inner class HeadLayoutHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private var mHeadAvatar: CircleImageView? = null
        private var mListenTime: TextView? = null
        private var mListenVip: TextView? = null
        private var mUserName: TextView? = null

        init {
            mHeadAvatar = itemView.findViewById(R.id.mine_head_avatar)
            mListenTime = itemView.findViewById(R.id.mine_head_listen_time)
            mListenVip = itemView.findViewById(R.id.mine_head_listen_vip)
            mUserName = itemView.findViewById(R.id.mine_head_user_name)
        }
    }

    internal inner class FuncLayoutHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        private var mLocalMusic: MineItemView? = null
        private var mDownMusic: MineItemView? = null
        private var mRecentMusic: MineItemView? = null
        private var mLoveMusic: MineItemView? = null
        private var mLoveSinger: MineItemView? = null
        private var mBuyMusic: MineItemView? = null

        init {
            mLocalMusic = itemView.findViewById(R.id.local_music)
            mDownMusic = itemView.findViewById(R.id.download_music)
            mRecentMusic = itemView.findViewById(R.id.recent_music)
            mLoveMusic = itemView.findViewById(R.id.love_music)
            mLoveSinger = itemView.findViewById(R.id.love_singer)
            mBuyMusic = itemView.findViewById(R.id.buy_music)

            mLocalMusic?.setOnClickListener(this)
            mDownMusic?.setOnClickListener(this)
            mRecentMusic?.setOnClickListener(this)
            mLoveMusic?.setOnClickListener(this)
            mLoveSinger?.setOnClickListener(this)
            mBuyMusic?.setOnClickListener(this)
        }

        fun onBind() {
            //            int singerCount = MusicUtils.queryArtist(mContext).size();
            //            int albumCount = MusicUtils.queryAlbums(mContext).size();
            //            int folderCount = MusicUtils.queryFolder(mContext).size();
            mLocalMusic!!.setmItemCount(MusicUtils.queryMusic(mContext, MConstants.START_FROM_LOCAL).size.toString())

        }

        override fun onClick(v: View?) {
            when (v!!.id) {
                R.id.local_music -> localMusic()
                R.id.download_music -> downloadMusic()
                R.id.recent_music -> recentMusic()
                R.id.love_music -> loveMusic()
                R.id.love_singer -> loveSinger()
                R.id.buy_music -> buyMusic()
            }
        }

        private fun localMusic() {
            val ft = (mContext as BaseActivity).supportFragmentManager.beginTransaction()
            ft.addToBackStack(null)
            ft.setCustomAnimations(R.anim.slide_right_in, R.anim.slide_left_out, R.anim.slide_left_in, R.anim.slide_right_out)
            ft.add(R.id.content, LocalMusicFragment.newInstance(0)).commit()
        }

        private fun downloadMusic() {
            val ft = (mContext as BaseActivity).supportFragmentManager.beginTransaction()
            ft.addToBackStack(null)
            ft.setCustomAnimations(R.anim.slide_right_in, R.anim.slide_left_out, R.anim.slide_left_in, R.anim.slide_right_out)
            ft.add(R.id.content, DownLoadFragment.newInstance()).commit()
        }

        private fun recentMusic() {
            val ft = (mContext as BaseActivity).supportFragmentManager.beginTransaction()
            ft.addToBackStack(null)
            ft.setCustomAnimations(R.anim.slide_right_in, R.anim.slide_left_out, R.anim.slide_left_in, R.anim.slide_right_out)
            ft.add(R.id.content, RecentMusicFragment.newInstance(0)).commit()
        }

        private fun loveMusic() {
            val intent = Intent(mContext, CollectedActivity::class.java)
            (mContext as BaseActivity).startActivityByX(intent, true)
        }

        private fun loveSinger() {
            val ft = (mContext as BaseActivity).supportFragmentManager.beginTransaction()
            ft.addToBackStack("KtUiLocalMusicFragment")
            ft.setCustomAnimations(R.anim.slide_right_in, R.anim.slide_left_out, R.anim.slide_left_in, R.anim.slide_right_out)
            ft.add(R.id.content, LocalMusicFragment.newInstance(0)).commit()
        }

        private fun buyMusic() {
            Toast.makeText(mContext, "测试", Toast.LENGTH_SHORT).show()
        }

    }

    internal inner class FavorTitleLayout(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private var mTitleLayout: RelativeLayout? = null
        private var mPlayListCount: TextView? = null

        init {
            mTitleLayout = itemView.findViewById(R.id.rl_favor_title)
            mPlayListCount = itemView.findViewById(R.id.playlist_count)
            mTitleLayout?.setOnClickListener(View.OnClickListener { favorTltle() })
        }


        private fun favorTltle() {
            val intent = Intent(mContext, SongListActivity::class.java)
            (mContext as BaseActivity).startActivityByX(intent, false)
        }
    }

    internal inner class ManageSongList(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            val intent = Intent(mContext, SongListActivity::class.java)
            (mContext as BaseActivity).startActivityByX(intent, false)
        }
    }


    internal inner class FavorItemLayout(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private var songListPic: SimpleDraweeView? = null
        private var mPlayListTitle: TextView? = null
        private var mPlayListInfo: TextView? = null
        private var mImgDown: ImageView? = null
        private var mItemLayout: RelativeLayout? = null

        private fun favorItem() {
            SongListInfoActivity.startActivity(mContext, mList!![adapterPosition - 3].getId(), mList!![adapterPosition - 3].getName())
        }

        init {
            songListPic = itemView.findViewById(R.id.img_album)
            mPlayListTitle = itemView.findViewById(R.id.play_list_title)
            mPlayListInfo = itemView.findViewById(R.id.play_list_info)
            mImgDown = itemView.findViewById(R.id.img_down)
            mItemLayout = itemView.findViewById(R.id.rl_favor_item)

            mItemLayout?.setOnClickListener(View.OnClickListener { favorItem() })
        }

        fun onBind(songListEntity: SongListEntity) {
            val musicEntity = MyApplication.musicInfoDBService.firstEntity(songListEntity.getId())
            if (musicEntity?.getAlbumPic() != null) {
                MyApplication.songListDBService.updatePic(songListEntity.getId(), musicEntity.getAlbumPic())
                songListPic!!.setImageURI(musicEntity.getAlbumPic())
            } else if (musicEntity != null) {
                if (MyApplication.imageDBService.query(musicEntity.getArtist().replace(";", "")) == null) {
                    val avatarRequest = AvatarRequest()
                    avatarRequest.artist = musicEntity.getArtist().replace(";", "")
                    MyOkHttpClient.getInstance(mContext).sendNet(avatarRequest, object : BaseCallback<AvatarResponse>() {
                        override fun onFailure(code: Int, error_msg: String) {

                        }

                        override fun onSuccess(statusCode: Int, response: AvatarResponse) {
                            MyApplication.imageDBService.insert(musicEntity.getArtist().replace(";", ""), response.artist.image[2].`_$Text112`)
                            MyApplication.songListDBService.updatePic(songListEntity.getId(), response.artist.image[3].`_$Text112`)
                            songListPic!!.setImageURI(response.artist.image[2].`_$Text112`)
                        }
                    })
                } else {
                    MyApplication.songListDBService.updatePic(songListEntity.getId(), MyApplication.imageDBService.query(musicEntity.getArtist().replace(";", "")))
                    songListPic!!.setImageURI(MyApplication.imageDBService.query(musicEntity.getArtist().replace(";", "")))
                }
            }
            mPlayListTitle!!.text = songListEntity.getName()
            mPlayListInfo!!.text = MyApplication.musicInfoDBService.getLocalCount(songListEntity.getId())
        }
    }

}